### PR TITLE
[GPU] Remove hand-rolled `static_assert`

### PR DESCRIPTION
**Pull request by FabianSchuetze**
The PR contains cosmetic changes to make the GPU code easier to read. The GPU code relies on template metaprogramming for a static check, but more concise c++11 functionality can replace these checks. Once PCL requires CUDA7 as the minimum CUDA version, we can remove the implementation of this static check altogether and replace the last remaining call with c++11 code.

### CODE REVIEW

**Comment by larshg** [comment]
https://github.com/PointCloudLibrary/pcl/blob/f8a48ddd61141601c6b718a4d15d9040ca0e1662/cmake/pcl_find_cuda.cmake#L9

Currently we search for version 9. Its also the minimum version on our CI. So I would say you can go ahead and remove it?

**Comment by FabianSchuetze** [comment]
Thank you so much, @larshg, for pointing this out. It always fills me with join if I can remove unneeded code. Thanks to your comment, we can even remove an entire file.

File: gpu/utils/include/pcl/gpu/utils/device/static_check.hpp [hunk 1]
```diff
@@ -1,69 +0,0 @@
...
-*  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
-*  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
-*  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
-*  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
-*  POSSIBILITY OF SUCH DAMAGE.
-*
-*  Author: Anatoly Baskeheev, Itseez Ltd, (myname.mysurname@mycompany.com)
-*/
-
-//TODO: Once PCL requires min CUDA7 (allows c++11 on device), remove file
```
Comment (kunaltyagi) [inline gpu/utils/include/pcl/gpu/utils/device/static_check.hpp]: Instead of removing, let's deprecate the header for PCL 1.15 release

**Review by kunaltyagi — CHANGES_REQUESTED** [review]

File: gpu/utils/include/pcl/gpu/utils/device/static_check.hpp (hunk 1 above)
Comment (FabianSchuetze) [reply gpu/utils/include/pcl/gpu/utils/device/static_check.hpp]: Thanks for your comment! I added a deprecated attribute. Did you had this in mind? Is there another way to mark pending deprecation?

File: gpu/utils/include/pcl/gpu/utils/device/static_check.hpp [hunk 2]
```diff
@@ -48,8 +48,10 @@ namespace pcl
     namespace device
     {
         template<bool expr> struct Static {};
-        
-        template<> struct Static<true> 
+
+        template <>
+        struct [[deprecated("This class will be replaced at PCL release 1.15  by "
```
Comment (kunaltyagi) [inline gpu/utils/include/pcl/gpu/utils/device/static_check.hpp:53]: Please also add a deprecation for the file itself. `PCL_DEPRECATED_HEADER`

**Comment by mvieth** [comment]
@FabianSchuetze Are you interested in continuing this pull request? Seems like it is almost done

File: gpu/utils/include/pcl/gpu/utils/device/static_check.hpp (hunk 2 above)
Comment (FabianSchuetze) [reply gpu/utils/include/pcl/gpu/utils/device/static_check.hpp]: Thanks for the comment, Kunal! I have added the deprecation notice. Is that what you had in mind?

**Comment by FabianSchuetze** [comment]
Thanks for your message, @mvieth . I am happy to continue with this PR and I tried to address Kunal's suggestion. The build failed twice, do you maybe have an idea how that is related to the code changes?

**Comment by larshg** [comment]
Seems like and odd failing unit test with following error:

 /Users/runner/work/1/s/test/search/test_organized.cpp:259: Failure
Expected: (pointDist) <= (searchRadius), actual: 0.275439 vs 0.275439

So I think we can just ignore it.

**Comment by larshg** [comment]
I'm getting this error:
error : expected a comma (the one-argument version of static_assert is not enabled in this mode)

Seems one-argument is first supported when c++17 is enabled on VS.

**Comment by mvieth** [comment]
I restarted the test and it should pass now, but yes, otherwise we can ignore it.
Oh right, in C++11 the static_assert also needs a message string. Could you add one each @FabianSchuetze ? Can be something simple.
https://en.cppreference.com/w/cpp/language/static_assert

**Comment by mvieth** [comment]
@FabianSchuetze Do you have time to continue this PR (add a simple message string for each static_assert)? Otherwise I might take over

File: gpu/features/src/features.cpp [hunk 3]
```diff
@@ -206,8 +205,8 @@ void pcl::gpu::PFHRGBEstimation::compute(DeviceArray2D<PFHRGBSignature250>& feat
 
 pcl::gpu::FPFHEstimation::FPFHEstimation()
 {
-    Static<sizeof(FPFHEstimation:: PointType) == sizeof(device:: PointType)>::check();
-    Static<sizeof(FPFHEstimation::NormalType) == sizeof(device::NormalType)>::check();    
+    static_assert(sizeof(FPFHEstimation:: PointType) == sizeof(device:: PointType));
+    static_assert(sizeof(FPFHEstimation::NormalType) == sizeof(device::NormalType));
```
Comment (mvieth) [inline gpu/features/src/features.cpp:209]:
```suggestion
    static_assert(sizeof(FPFHEstimation:: PointType) == sizeof(device:: PointType), "Point sizes do not match");
    static_assert(sizeof(FPFHEstimation::NormalType) == sizeof(device::NormalType), "Normal sizes do not match");
```

File: gpu/features/src/features.cpp [hunk 4]
```diff
@@ -298,8 +297,8 @@ void pcl::gpu::PPFEstimation::compute(DeviceArray<PPFSignature>& features)
 
 void pcl::gpu::PPFRGBEstimation::compute(DeviceArray<PPFRGBSignature>& features)
 {    
-    Static<sizeof(PPFEstimation:: PointType) == sizeof(device:: PointType)>::check();
-    Static<sizeof(PPFEstimation::NormalType) == sizeof(device::NormalType)>::check();    
+    static_assert(sizeof(PPFEstimation:: PointType) == sizeof(device:: PointType));
+    static_assert(sizeof(PPFEstimation::NormalType) == sizeof(device::NormalType));
```
Comment (mvieth) [inline gpu/features/src/features.cpp:301]:
```suggestion
    static_assert(sizeof(PPFEstimation:: PointType) == sizeof(device:: PointType), "Point sizes do not match");
    static_assert(sizeof(PPFEstimation::NormalType) == sizeof(device::NormalType), "Normal sizes do not match");
```

File: gpu/features/src/features.cpp [hunk 5]
```diff
@@ -341,8 +340,8 @@ void pcl::gpu::PPFRGBRegionEstimation::compute(DeviceArray<PPFRGBSignature>& fea
 
 void pcl::gpu::PrincipalCurvaturesEstimation::compute(DeviceArray<PrincipalCurvatures>& features)
 {
-    Static<sizeof(PPFRGBRegionEstimation:: PointType) == sizeof(device:: PointType)>::check();
-    Static<sizeof(PPFRGBRegionEstimation::NormalType) == sizeof(device::NormalType)>::check();    
+    static_assert(sizeof(PPFRGBRegionEstimation:: PointType) == sizeof(device:: PointType));
+    static_assert(sizeof(PPFRGBRegionEstimation::NormalType) == sizeof(device::NormalType));
```
Comment (mvieth) [inline gpu/features/src/features.cpp:344]:
```suggestion
    static_assert(sizeof(PPFRGBRegionEstimation:: PointType) == sizeof(device:: PointType), "Point sizes do not match");
    static_assert(sizeof(PPFRGBRegionEstimation::NormalType) == sizeof(device::NormalType), "Normal sizes do not match");
```

**Review by kunaltyagi — APPROVED** [review]
🚀

**Review by mvieth — APPROVED** [review]

**Review by larshg — APPROVED** [review]

**Pull request merged**